### PR TITLE
[dtensor] refactor op dispatch and fix is_same_size/equal

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112931
* __->__ #112927

torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

This PR refactors the existing op_dispatch to make it a class instance
so that we can do custom op handling, then fixes both torch.equal and
torch.is_same_size

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

### CODE REVIEW

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 1]
```diff
@@ -59,6 +57,45 @@ def default_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
...
+        self_strategy
+        if self_strategy.max_num_shards() >= other_strategy.max_num_shards()
+        else other_strategy
+    )
+    equal_strategy = OpStrategy([])
+
+    for arg_strategy in select_strategy.strategies:
+        arg_spec = arg_strategy.output_spec
+        if is_tensor_partial(arg_spec):
+            # if the arg_spec have partial, reshard to replicate
```
Comment (wconstab) [inline torch/distributed/_tensor/ops/tensor_ops.py:85]: expand the comment to say 'why' partial ==> replicate?

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 2]
```diff
@@ -59,6 +57,45 @@ def default_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
...
 
 
+@register_op_strategy(
+    [
+        aten.equal.default,
+        aten.is_same_size.default,
+    ]
+)
+def equal_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
+    # equal_strategy deals with ops that comparing two tensor, we need to make sure
```
Comment (wconstab) [inline torch/distributed/_tensor/ops/tensor_ops.py:67]: is the idea here that before you can perform an equality test on 2 ops, you need 
 - (a) they have the same logical shape/size- otherwise can skip comparing data
 - (b) if a dim is sharded for one arg, but replicated for another arg, its nots straghtforward to memcmp their content locally
 - (c) so therefore we canonicalize the args by making them all replicated, so they are easy to compare locally


if so, then i wonder if its more optimal to consider for (b) if both input args are sharded, but already have same sharding, could you skip the replicate step and do the comparison on a local-shard basis?

**Review by wconstab — COMMENTED** [review]
some questions mostly for my learning. maybe worth having someone else stamp but overall i think it looks right

File: torch/distributed/_tensor/op_schema.py [hunk 3]
```diff
@@ -257,6 +257,11 @@ def return_type_tuple_tensors(self) -> bool:
             return_types[0].type, torch.TensorType
         )
 
+    def return_type_tensor(self) -> bool:
+        return_types = self.op._schema.returns
+        # all dispatch ops only return Tensor or Tuple[Tensor], so this check if enough
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_tensor/op_schema.py:262]: Oh yes I should clarify in the comments: op only return tensor or tuple/list of tensors for tensor like return types, some op do return int/bool, therefore I need to check if the return type is a tensor type or not.

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 4]
```diff
@@ -59,6 +57,45 @@ def default_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
     )
 
 
+@register_op_strategy(
+    [
+        aten.equal.default,
+        aten.is_same_size.default,
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_tensor/ops/tensor_ops.py:63]: yep is_same_size would never hit. I mainly leave it here because if logically these two share the same logic if we don't do any dispatching optimizations in this PR, and for completeness I was mostly thinking if we plan to explore "single graph expansion like dist compiler" in the future again, having all ops registered to one table might be better, but I can remove it if you feel we want to keep things consistent ;)

Because for `is_same_size` we don't need to run propagation as we can just directly compare the DTensor shapes, the dispatching logic is more like a optimization.

[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/_tensor/ops/tensor_ops.py:63]: ok that convinced me, this is fine

File: test/distributed/_tensor/test_tensor_ops.py [hunk 5]
```diff
@@ -253,6 +253,17 @@ def test_equal(self):
...
+        self.assertTrue(dist_tensor_1.is_same_size(dist_tensor_2))
+
+        # test if sharding are different
+        replica_spec = [Replicate()]
+        global_input = torch.ones(4 * self.world_size, 4)
+        dist_tensor_3 = DTensor.from_local(
+            global_input, device_mesh, replica_spec, run_check=False
+        )
+
+        self.assertTrue(dist_tensor_1.equal(dist_tensor_3))
```
Comment (wconstab) [inline test/distributed/_tensor/test_tensor_ops.py:265]: should we add tests for cases where values actually are different, and especially, for the corner case where the difference only occurs on a remote shard and we notice it by doing the replication?

File: torch/distributed/_tensor/ops/tensor_ops.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_tensor/ops/tensor_ops.py:67]: Yep this is a great summary, I'll put these into comments too!

> (a) they have the same logical shape/size- otherwise can skip comparing data
Yep this is a good point we can do such a optimization similar to is_same_size! but only compare the data when logical shape are the same

> i wonder if its more optimal to consider for (b) if both input args are sharded, but already have same sharding, could you skip the replicate step and do the comparison on a local-shard basis?

Yes this is what we are doing currently, currently I'm performing this check inside [sharding propagation](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/sharding_prop.py#L200), if all inputs are sharded in the same way, we don't need to do resharding to turn them as replicated. This strategy mainly try to follow one argument's sharding with the max number of shards, and see if every arg have the same sharding layout, if so, directly compare, otherwise, redistribute to the argument's sharding first.

File: torch/distributed/_tensor/dispatch.py [hunk 6]
```diff
@@ -125,6 +125,12 @@ def _operator_dispatch(
     kwargs: Dict[str, object],
     sharding_propagator: ShardingPropagator,
 ) -> Tuple[object, OpSchema, OutputSharding]:
+    # operators that does not need to go through sharding propagation
+    if op_call == aten.is_same_size.default:
```
Comment (fduwjj) [inline torch/distributed/_tensor/dispatch.py:129]: maybe not in this PR. But down the road, IMHO, we want to avoid this type check(for single or certain ops) as much as possible?

**Review by fduwjj — APPROVED** [review]
Unblock and please address Will's comment.

File: torch/distributed/_tensor/ops/tensor_ops.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/ops/tensor_ops.py:63]: Do you think it's a good idea to keep all lists of ops (e.g. `skip_prop_list`, `random_op_list`) in one file for better readability?

**Review by XilunWu — APPROVED** [review]
LGTM. We can address Will's comment in a follow up PR.

File: test/distributed/_tensor/test_tensor_ops.py (hunk 5 above)
Comment (wanchaol) [reply test/distributed/_tensor/test_tensor_ops.py:265]: added.

File: torch/distributed/_tensor/dispatch.py (hunk 6 above)
Comment (wanchaol) [reply torch/distributed/_tensor/dispatch.py]: I refactored the dispatch code a bit to not have the op type checks, now all things are handled by the `custom_op_handlers`

**Comment by wanchaol** [comment]
> LGTM. We can address Will's comment in a follow up PR.

I think I already addressed them so it's good :)
